File: algo_2021/w4.-queue-bfs/problems/1179.Friend_Circles/1179_friend_circles__bfs.java
public class Solution {
    /**
     * @param M: a matrix
     * @return: the total number of friend circles among all the students
     */
    public int findCircleNum(int[][] M) {
        // Write your code here
        Queue<Integer> queue = new LinkedList<Integer>();
        boolean[] visited = new boolean[M.length];
        int cnt = 0;

        for (int roll = 0; roll < M.length; roll++) {
            if (visited[roll] == true) 
                continue;
            queue.offer(roll);
            while (!queue.isEmpty()) {
                int poll = queue.poll();
                for (int col = 0; col < M[poll].length; col++) {
                    if ((M[poll][col] != 1) || (visited[col] == true)) 
                        continue;
                    queue.offer(col);
                    visited[col] = true;
                }
            }
            cnt++;
        }
        return cnt;
    }
}

File: algo_2021/s4.-queue-bfs-graph/1179.Friend_Circles/1179_friend_circles__bfs.py
class Solution:
    """
    @param M: a matrix
    @return: the total number of friend circles among all the students
    """
    def findCircleNum(self, M):
        # Write your code here
        queue = collections.deque()
        visited = [False] * len(M)
        cnt = 0
        
        # BFS, search by level
        # start traverse roll 
        for roll in range(len(M)):
            # if row already searched, skip
            if visited[roll]:
                continue
            queue.append(roll)
            # if queue not empty, means still have connection within this group
            while queue:
                poll = queue.popleft()
                # start traverse in column
                for col in range(len(M[poll])):
                    # if that column value is not 1, or already visited, skip
                    if M[poll][col] != 1 or visited[col]:
                        continue
                    queue.append(col)
                    visited[col] = True
            # after entire while loop, mean this group's all connections have been searched
            cnt+=1
        return cnt        